scipy/optimize/tests/test__differential_evolution.py

```diff
@@ -1236,3 +1237,41 @@ def c1(x):
         assert_(np.all(np.array(c1(res.x)) <= 0.001))
         assert_(np.all(res.x >= np.array(bounds)[:, 0]))
         assert_(np.all(res.x <= np.array(bounds)[:, 1]))
+
+    def test_integrality(self):
+        # test fitting discrete distribution to data
+        rng = np.random.default_rng(6519843218105)
+        dist = stats.nbinom
+        shapes = (5, 0.5)
+        x = dist.rvs(*shapes, size=10000, random_state=rng)
+
+        def func(free_params, *args):
+            dist, x = args
+            # negative log-likelihood function
+            ll = -np.log(dist.pmf(x, *free_params)).sum(axis=-1)
+            if np.isnan(ll):  # occurs when x is outside of support
+                ll = np.inf  # we don't want that
+            return ll
+
+        integrality = [True, False]
+        bounds = [(1, 18), (0, 0.95)]
+
+        res = differential_evolution(func, bounds, args=(dist, x),
+                                     integrality=integrality, polish=False,
+                                     seed=rng)
+        # tolerance has to be fairly relaxed for the second parameter
+        # because of the number of rvs samples generated. The more you generate
+        # the closer the parameter will come to the theoretical value (it will
+        # never be the theoretical value because they're random variates).
+        assert res.x[0] == 5
+        assert_allclose(res.x, shapes, rtol=0.02)
+
```

Comment:
This is great that it comes close!
Later, we can enable polishing with the integral values fixed, and this would get nice and close.